xgboost/xgboost_cv_terminator.py

```diff
@@ -0,0 +1,63 @@
+"""
+Optuna example that demonstrates a pruner for XGBoost.cv.
+
+In this example, we optimize the validation auc of cancer detection using XGBoost.
+We optimize both the choice of booster model and their hyperparameters. Throughout
+training of models, a terminator observes trials and stops studies after insignificant
+improvement.
+
+You can run this example as follows:
+    $ python xgboost_cv_terminator.py
+
+"""
+
+import optuna
+from optuna.terminator import report_cross_validation_scores
+from optuna.terminator import TerminatorCallback
+
+import sklearn.datasets
+import xgboost as xgb
+
+
+def objective(trial):
+    train_x, train_y = sklearn.datasets.load_breast_cancer(return_X_y=True)
+    dtrain = xgb.DMatrix(train_x, label=train_y)
+
+    param = {
+        "verbosity": 0,
+        "objective": "binary:logistic",
+        "eval_metric": "auc",
+        "booster": "gbtree",
+        "lambda": trial.suggest_float("lambda", 1e-8, 1.0, log=True),
+        "alpha": trial.suggest_float("alpha", 1e-8, 1.0, log=True),
+    }
+
+    param["max_depth"] = trial.suggest_int("max_depth", 1, 9)
+    param["eta"] = trial.suggest_float("eta", 1e-8, 1.0, log=True)
+    param["gamma"] = trial.suggest_float("gamma", 1e-8, 1.0, log=True)
+    param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])
+
+    history = xgb.cv(param, dtrain, num_boost_round=100)
+
+    report_cross_validation_scores(trial, history["test-auc-mean"])
```

Comment:
In this line of the code, the mean values are passed to `report_cross_validation_scores`. However, it actually expects the individual AUC values.

Reply:
I roughly read the `xgboost` code, and I couldn't find a straightforward (i.e., arguments of the `xgb.cv` function) way to get the individual values of AUC. Do you have any ideas?

https://github.com/dmlc/xgboost/blob/2718ff530ca9a04023ea147424874e686762a384/python-package/xgboost/training.py#L577-L584

Reply:
The only thing I can think of is verbose_eval=1. Sklearn cv might work. I'm not sure, though. I will have to test it

Reply:
No, I think verbose eval with cv also returns the means

Reply:
I checked the history object using pdb and found that individual values of AUC are not available.

```
(Pdb) p history
    train-auc-mean  train-auc-std  test-auc-mean  test-auc-std
0         0.986280       0.004890       0.957351      0.009828
1         0.986408       0.004781       0.955515      0.009408
2         0.987039       0.003988       0.957704      0.012139
3         0.987009       0.003995       0.958738      0.011493
4         0.986989       0.003988       0.957905      0.012402
..             ...            ...            ...           ...
95        0.987098       0.003941       0.958840      0.011693
96        0.987084       0.003961       0.958840      0.011693
97        0.987098       0.003941       0.958860      0.011720
98        0.987084       0.003961       0.958840      0.011693
99        0.987098       0.003941       0.958840      0.011693

[100 rows x 4 columns]
```

It appears that `report_cross_validation_scores` is not available with `xgboost.cv()`. One possible approach is to write the cross-validation logic ourselves, but this would be complex and may not be suitable as an example code. Therefore, could we consider closing this pull request?

@nicholasptran Thanks for your huge effort :bow: What do you think?

Reply:
Sounds good to me